### PR TITLE
Run multi-tenant queries in parallel.

### PR DESCRIPTION
**Pull request by jeschkies**
**What this PR does / why we need it**:
Previously the multi-tenant queries would run in sequence which might have slowed them down. That is why we are going to execute them in a small worker pool.

The tests have been update as well, since they were faulty.

**Checklist**
- [x] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] Documentation added
- [x] Tests updated
- [ ] `CHANGELOG.md` updated
  - [ ] If the change is worth mentioning in the release notes, add `add-to-release-notes` label
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/setup/upgrade/_index.md`
- [ ] For Helm chart changes bump the Helm chart version in `production/helm/loki/Chart.yaml` and update `production/helm/loki/CHANGELOG.md` and `production/helm/loki/README.md`. [Example PR](https://github.com/grafana/loki/commit/d10549e3ece02120974929894ee333d07755d213)

### CODE REVIEW

File: pkg/storage/stores/series/series_index_store.go [hunk 1]
```diff
@@ -279,28 +278,25 @@ func (c *indexReaderWriter) chunksToSeries(ctx context.Context, in []logproto.Ch
...
-	}
-
-	if err := concurrency.ForEachJob(
-		ctx,
-		len(jobs),
-		numConcurrent,
-		func(_ context.Context, idx int) error {
-			res, err := jobs[idx]()
-			if res != nil {
-				results = append(results, res...)
```
Comment (jeschkies) [inline pkg/storage/stores/series/series_index_store.go:297]: I don't think this was thread safe so I've replace the logic.

File: pkg/querier/multi_tenant_querier.go [hunk 2]
```diff
@@ -84,17 +86,18 @@ func (q *MultiTenantQuerier) SelectSamples(ctx context.Context, params logql.Sel
...
-		singleContext := user.InjectOrgID(ctx, id)
-		iter, err := q.Querier.SelectSamples(singleContext, params)
-		if err != nil {
-			return nil, err
-		}
-
-		iters[i] = NewTenantSampleIterator(iter, id)
-		i++
+		id := id
+		p.Go(func(ctx context.Context) (iter.SampleIterator, error) {
```
Comment (jeschkies) [inline pkg/querier/multi_tenant_querier.go:92]: `iter.MapErr` would be a little nicer but it does not work with a context and cancel on error. There's one (https://github.com/sourcegraph/conc/pull/114/) in the work.

File: pkg/querier/multi_tenant_querier.go [hunk 3]
```diff
@@ -84,17 +86,18 @@ func (q *MultiTenantQuerier) SelectSamples(ctx context.Context, params logql.Sel
...
-		if err != nil {
-			return nil, err
-		}
-
-		iters[i] = NewTenantSampleIterator(iter, id)
-		i++
+		id := id
+		p.Go(func(c context.Context) (iter.SampleIterator, error) {
+			singleContext := user.InjectOrgID(c, id)
+			iter, err := q.Querier.SelectSamples(singleContext, params)
```
Comment (jeschkies) [inline pkg/querier/multi_tenant_querier.go:94]: This fails with a cancelled context when run a second time. I have a hard time understanding why.

**Comment by dannykopping** [comment]
@jeschkies what shall we do with this?

**Comment by dannykopping** [comment]
Closing for now, we can revive it later if we want to.

**Pull request closed without merging**